src/centralized.rs

```diff
@@ -0,0 +1,499 @@
+    n_p: usize,
+}
+
+impl Round {
+    /// Oracle producing a uniformly random value in [1, n_p] used for round candidates
+    /// We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h1(data: Vec<Vec<u8>>, n_p: usize) -> (Hash, usize) {
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(data);
+        (digest, utils::oracle_uniform(&digest, n_p))
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub fn new(v: usize, t: usize, n_p: usize) -> Round {
+        let mut data = vec![v.to_ne_bytes().to_vec()];
+        data.push(t.to_ne_bytes().to_vec());
+        let (h, h_usize) = Round::h1(data, n_p);
+        Round {
+            v,
+            t,
+            s_list: vec![],
+            h,
+            h_usize,
+            n_p,
+        }
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Round, s: Data) -> Round {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let mut data = vec![r.h.clone().to_vec()];
+        data.push(s.to_vec());
+        let (h, h_usize) = Round::h1(data, r.n_p);
+        Round {
+            v: r.v,
+            t: r.t,
+            s_list,
+            h,
+            h_usize,
+            n_p: r.n_p,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
+    /// Proof counter
+    r: usize,
+    /// Proof 2nd counter
+    d: usize,
+    /// Proof tuple
+    items: Vec<Data>,
+}
+
+impl Proof {
+    /// Returns a new proof
+    fn new() -> Self {
+        Proof {
+            r: 0,
+            d: 0,
+            items: vec![],
+        }
+    }
+
+    /// Oracle producing a uniformly random value in [1, n_p] used for prehashing S_p
+    fn h0(setup: &Setup, v: usize, s: Data) -> usize {
+        let mut data = vec![v.to_ne_bytes().to_vec()];
+        data.push(s.to_vec());
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(data);
+        utils::oracle_uniform(&digest, setup.n_p)
+    }
+
+    /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0 otherwise
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut data = vec![];
+        data.push(r.v.to_ne_bytes().to_vec());
+        data.push(r.t.to_ne_bytes().to_vec());
+        for s in &r.s_list {
+            data.push(s.clone().to_vec());
+        }
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(data);
+        utils::oracle_binomial(&digest, setup.q)
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &Vec<Vec<Data>>,
```

Comment:
In Rust, it is preferable to use references to a slice (`&[Vec<Data>]`) rather than `&Vec<Vec<Data>>` unless you are explicitly interacting with some of the features of `Vec` (i.e., pushing or resizing). This is slightly more performant since you are not tying your parameter via the vector (allocated on the heap) but only to a contiguous portion of it. You are therefore avoiding some of the unnecessary heap management related to handling a `Vec`.  This gives you more flexibility too since you can pass a slice of an array to the function.

Reply:
I hear you.

The problem I have with this is that I would have to initialize the array with setup.n_p and as it is not constant, it is not (easily perhaps) possible. 

Another problem is that Vector does not implement the Copy trait, as such to initialise bins I would have to do something like: `let bins : [Vec<Element>; N] = vec![vec![]; N].try_into().unwrap();`